File: examples/x3_oop_bullet.py
#!/usr/bin/env python3
'''
@author Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

import g2d, random
from actor import Actor, Arena

class Alien(Actor):
    def __init__(self, arena, pos):
        self._x, self._y = pos
        self._w, self._h = 20, 20
        self._xmin, self._xmax = x0, x0 + 150
        self._dx, self._dy = 5, 5
        self._arena = arena
        arena.add(self)

    def move(self):
        if self._xmin <= self._x + self._dx <= self._xmax:
            self._x += self._dx
        else:
            self._dx = -self._dx
            self._y += self._dy

    def position(self):
        return self._x, self._y, self._w, self._h

    def symbol(self):
        return 0, 0

    def collide(self, other):
        pass

class Bullet(Actor):
    def __init__(self, arena, x0: int):
        self._w, self._h = 5, 10
        self._x, self._y = x0, arena.size()[1] - self._h
        self._dy = -5
        self._arena = arena
        arena.add(self)

    def move(self):
        self._y += self._dy
        if self._y < 0:
            self._arena.remove(self)

    def position(self):
        return self._x, self._y, self._w, self._h

    def symbol(self):
        return 0, 0

    def collide(self, other):
        if isinstance(other, Alien):
            self._arena.remove(other)
            self._arena.remove(self)

def tick():
    g2d.clear_canvas()
    if random.randrange(50) == 0:
        Bullet(arena, random.randrange(arena.size()[0]))
    arena.move_all()
    for a in arena.actors():
        g2d.fill_rect(a.position())

def main():
    global arena
    arena = Arena((320, 240))
    Alien(arena, (40, 40))
    Alien(arena, (80, 80))
    g2d.init_canvas(arena.size())
    g2d.main_loop(tick)

main()

File: go/x3_oop_bullet.go
package main

import . "g2d"

var arena = NewArena(Point{480, 360})
var a1 = NewAlien(arena, Point{40, 40})
var a2 = NewAlien(arena, Point{80, 80})

type Alien struct {
    arena      *Arena
    x, y, w, h int
    xmin, xmax int
    dx, dy     int
}

func NewAlien(arena *Arena, pos Point) *Alien {
    a := &Alien{arena, pos.X, pos.Y, 20, 20, pos.X, pos.X+150, 5, 5}
    arena.Add(a)
    return a
}

func (a *Alien) Move() {
    if a.xmin <= a.x+a.dx && a.x+a.dx <= a.xmax {
        a.x += a.dx
    } else {
        a.dx = -a.dx
        a.y += a.dy
    }
}

func (a *Alien) Position() Point {
    return Point{a.x, a.y}
}

func (a *Alien) Size() Point {
    return Point{a.w, a.h}
}

func (a *Alien) Symbol() Point {
    return Point{0, 0}
}

func (a *Alien) Collide(other Actor) {
}

type Bullet struct {
    arena          *Arena
    x, y, w, h, dy int
}

func NewBullet(arena *Arena, x int) *Bullet {
    a := &Bullet{arena, x, arena.Size().Y - 10, 5, 10, -5}
    arena.Add(a)
    return a
}

func (a *Bullet) Move() {
    a.y += a.dy
    if a.y < 0 {
        a.arena.Remove(a)
    }
}

func (a *Bullet) Position() Point {
    return Point{a.x, a.y}
}

func (a *Bullet) Size() Point {
    return Point{a.w, a.h}
}

func (a *Bullet) Symbol() Point {
    return Point{0, 0}
}

func (a *Bullet) Collide(other Actor) {
    if _, ok := other.(*Alien); ok {
        a.arena.Remove(other)
        a.arena.Remove(a)
    }
}

func tick() {
    if RandInt(0, 49) == 0 {
        NewBullet(arena, RandInt(0, arena.Size().X))
    }
    ClearCanvas()
    arena.MoveAll()
    for _, actor := range arena.Actors() {
        FillRect(actor.Position(), actor.Size())
    }
}

func main() {
    InitCanvas(arena.Size())
    MainLoop(tick)
}
